sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -329,6 +334,48 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     }
   }
 
+  private def unwrapTimeStampToDate(
+      exp: BinaryComparison,
+      fromExp: Expression,
+      ts: Literal,
+      tz: Option[String],
+      evalMode: EvalMode.Value): Expression = {
+    val floorDate = Cast(ts, fromExp.dataType, tz, evalMode)
+    val dateAddOne = DateAdd(floorDate, Literal(1, IntegerType))
+    val isStartOfDay =
+      EqualTo(ts, Cast(floorDate, ts.dataType, tz, evalMode)).eval(EmptyRow).asInstanceOf[Boolean]
+
+    exp match {
+      case _: GreaterThan =>
+        GreaterThan(fromExp, floorDate)
+      case _: GreaterThanOrEqual =>
+        if (isStartOfDay) {
+          GreaterThanOrEqual(fromExp, floorDate)
+        } else {
+          GreaterThanOrEqual(fromExp, dateAddOne)
+        }
+      case _: EqualTo =>
+        if (isStartOfDay) {
+          EqualTo(fromExp, floorDate)
+        } else if (!fromExp.nullable) {
+          FalseLiteral
+        } else {
+          And(EqualTo(fromExp, floorDate), EqualTo(fromExp, dateAddOne))
```

Comment:
This is very confusing. How can the `fromExp` be equal to two different values at the same time? This expression is either false or null.

Do you mean `if(isnull(e), null, false)`? Then `And(IsNull(e), Literal(null, BooleanType))` should be more efficient.

Reply:
actually there is already a `falseIfNotNull` method in this class.

Reply:
BTW let's keep the high code review quality in the community. This is the core logic of this optimization and it's even highlighted in the PR description. I'm surprised we missed this case.